src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java

```diff
@@ -0,0 +1,82 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.randomized;
+
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizationException;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.QueryOptimizationContext;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RuleInstances;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.utils.CostEstimationUtils;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.utils.PhysicalPlanWithCost;
+
+public class SimulatedAnnealing extends RandomizedQueryOptimizerBase {
+
+	public SimulatedAnnealing(	final QueryOptimizationContext context,
+								final RuleInstances rewritingRules) {
+		super(context, rewritingRules);
+	}
+
+	@Override
+	public PhysicalPlan optimize(PhysicalPlan initialPlan) throws QueryOptimizationException {
+		
+		// The first plan, which is currently the best plan we know of.
+		PhysicalPlanWithCost bestPlan = new PhysicalPlanWithCost(initialPlan, CostEstimationUtils.getEstimates( context.getCostModel(), initialPlan )[0]);
+		PhysicalPlanWithCost currentPlan = bestPlan;
+		
+		// A temperature value is assigned based on the cost. Because of how this value is used, the only important thing is its size relative to the cost of the current plan.
+		double temperature = currentPlan.getWeight()*2;
+		int unchanged = 0;
+		
+		// Possible TODO: Define what frozen means. Whether to hardcode a temp threshhold below which is frozen, or to pass on a frozen threshhold to the constructor, or something else entirely.
+		// I'm using the paper's implementation of frozen for now.
+		// While not frozen, do
+		while(!isFrozen(temperature,unchanged)) {
+			
+			unchanged += 1;
+			
+			// TODO: Define what equilibrium entails on a practical level
+			// The paper's implementation counts equilibrium as 16*(number of joins in query).
+			while(!isEquilibrium()) {
+			
+				// Here we want a random neighbour, and its cost
+				// My first instinct to do something like
+				// final PhysicalPlanWithCost alternativePlan = (PhysicalPlan alternativePlanPlan = getRandom(getNeighbours(currentPlan.getPlan())), CostEstimationUtils.getEstimates( context.getCostModel(), alternativePlanPlan )[0])
+				// to make the code really compact.
+				// because it seems wasteful to create a local variable that's stored in optimize() just because we need to get the cost separately.
+				final PhysicalPlan temporaryPlan = getRandom(getNeighbours(currentPlan.getPlan()));
+				final PhysicalPlanWithCost alternativePlan = new PhysicalPlanWithCost(temporaryPlan, CostEstimationUtils.getEstimates( context.getCostModel(), temporaryPlan )[0]);
+				// Alternatively, a better option is possibly to create a new constructor for PhysicalPlanWithCost...
+				// ...that only takes a PhysicalPlan argument, and gets the cost itself.
```

Comment:
No, don't create such a new constructor -- see [my earlier comment](https://github.com/LiUSemWeb/HeFQUIN/pull/134#issuecomment-991719903).

However, here's what you can do: Please add the following method to `PhysicalPlanWithCostUtils`:
```java
    public static PhysicalPlanWithCost annotatePlanWithCost( final CostModel cm, final PhysicalPlan plan ) throws QueryOptimizationException {
        final Double[] costs;
        try {
            costs = CostEstimationUtils.getEstimates(cm, plan);
        } catch ( final CostEstimationException e ) {
            throw new QueryOptimizationException( "Determining the cost for the plan caused an exception.", e.getCause() );
        }

        return new PhysicalPlanWithCost( plan, costs[0] );
    }
```
Once you have that, you can change the line here as follows:
```suggestion
				final PhysicalPlanWithCost alternativePlan = PhysicalPlanWithCostUtils.annotatePlanWithCost( context.getCostModel(), temporaryPlan );
```

Reply:
In fact, you can then also use this new helper function in line 21 above.